File: typescript/tests/core/compose-direct.test.ts
import { describe, expect, it } from 'vitest';
import { PetriNet } from '../../src/core/petri-net.js';
import { place } from '../../src/core/place.js';
import { SubnetDef } from '../../src/core/subnet-def.js';
import { Transition } from '../../src/core/transition.js';
import { one } from '../../src/core/in.js';
import { outPlace } from '../../src/core/out.js';
import { FusionSet } from '../../src/core/fusion-set.js';
import { TimePetriNetAnalyzer, type LivenessResult } from '../../src/verification/analysis/time-petri-net-analyzer.js';
import { MarkingState } from '../../src/verification/marking-state.js';
import { producer, consumer, retryPolicy, leakyBucket } from '../fixtures/subnet-fixtures.js';
import { bindProducers } from '../fixtures/producing-actions.js';

/**
 * Tests for `PetriNetBuilder.compose(def)` — direct composition per
 * **MOD-025**.
 *
 * Rule: a `SubnetDef` composed directly is integrated *without*
 * instantiation/prefix-renaming. Body places and transitions keep their
 * original names; places merge into the host by name. Direct composition is
 * order-independent.
 *
 * Mirrors `java/src/test/java/org/libpetri/core/ComposeDirectTest.java`
 * (scenario 9 — place token-type conflict — is Java-only: TS Place equality
 * is name-only at runtime, see MOD-025).
 */
describe('PetriNetBuilder.compose(def) — MOD-025 direct composition', () => {
  /** A producer whose output place is named `pipe` (wires by name). */
  function pipeProducer(): SubnetDef<void> {
    const seed = place<string>('seed');
    const pipe = place<string>('pipe');
    return SubnetDef.builder('PipeProducer')
      .place(seed)
      .place(pipe)
      .transition(Transition.builder('emit').inputs(one(seed)).outputs(outPlace(pipe)).build())
      .build();
  }

  /** A consumer whose input place is named `pipe` (wires by name). */
  function pipeConsumer(): SubnetDef<void> {
    const pipe = place<string>('pipe');
    const sink = place<string>('sink');
    return SubnetDef.builder('PipeConsumer')
      .place(pipe)
      .place(sink)
      .transition(Transition.builder('eat').inputs(one(pipe)).outputs(outPlace(sink)).build())
      .build();
  }

  const names = (xs: Iterable<{ name: string }>): string[] => [...xs].map((x) => x.name).sort();

  it('composeDirect_mergesBodyPlacesByName', () => {
    const hostOutput = place<string>('output');
    const host = PetriNet.builder('Host').place(hostOutput).compose(producer()).build();

    expect([...host.places].filter((p) => p.name === 'output')).toHaveLength(1);
    expect([...host.places].some((p) => p.name === 'nextItem')).toBe(true);

    const produce = [...host.transitions].find((t) => t.name === 'produce')!;
    expect([...produce.outputPlaces()].some((p) => p.name === 'output')).toBe(true);
  });

  it('composeDirect_hostHasNoPreDeclaredPlace_arrivesUnPrefixed', () => {
    const host = PetriNet.builder('Host').compose(producer()).build();
    expect([...host.places].some((p) => p.name === 'output')).toBe(true);
    expect([...host.places].some((p) => p.name === 'nextItem')).toBe(true);
  });

  it('composeDirect_introducesNoPrefixSeparator', () => {
    const host = PetriNet.builder('Host').compose(producer()).compose(consumer()).build();
    expect([...host.places].some((p) => p.name.includes('/'))).toBe(false);
    expect([...host.transitions].some((t) => t.name.includes('/'))).toBe(false);
  });

  it('composeDirect_isOrderIndependent', () => {
    const prodThenCons = PetriNet.builder('Host').compose(pipeProducer()).compose(pipeConsumer()).build();
    const consThenProd = PetriNet.builder('Host').compose(pipeConsumer()).compose(pipeProducer()).build();

    expect(names(prodThenCons.places)).toEqual(names(consThenProd.places));
    expect(names(prodThenCons.transitions)).toEqual(names(consThenProd.transitions));

    for (const tName of ['emit', 'eat']) {
      const a = [...prodThenCons.transitions].find((t) => t.name === tName)!;
      const b = [...consThenProd.transitions].find((t) => t.name === tName)!;
      expect(names([...a.inputSpecs].map((s) => s.place))).toEqual(names([...b.inputSpecs].map((s) => s.place)));
      expect(names(a.outputPlaces())).toEqual(names(b.outputPlaces()));
    }
  });

  it('composeDirect_structurallyEqualToHandWrittenFlatNet', () => {
    const composed = PetriNet.builder('Net').compose(pipeProducer()).compose(pipeConsumer()).build();

    const seed = place<string>('seed');
    const pipe = place<string>('pipe');
    const sink = place<string>('sink');
    const handWritten = PetriNet.builder('Net')
      .transition(Transition.builder('emit').inputs(one(seed)).outputs(outPlace(pipe)).build())
      .transition(Transition.builder('eat').inputs(one(pipe)).outputs(outPlace(sink)).build())
      .build();

    expect(names(composed.places)).toEqual(names(handWritten.places));
    expect(names(composed.transitions)).toEqual(names(handWritten.transitions));
  });

  it('composeDirect_mixedWithInstanceCompose_coexist', () => {
    const host = PetriNet.builder('Host')
      .compose(producer())
      .compose(consumer().instantiate('c1'))
      .build();

    const tNames = [...host.transitions].map((t) => t.name);
    expect(tNames).toContain('produce');     // direct — un-prefixed
    expect(tNames).toContain('c1/consume');  // instance — prefixed
  });

  it('composeDirect_transitionNameCollision_throws', () => {
    const builder = PetriNet.builder('Host').compose(producer());
    let err: Error | undefined;
    try {
      builder.compose(producer());
    } catch (e) {
      err = e as Error;
    }
    expect(err).toBeDefined();
    expect(err!.message).toContain('produce');
    expect(err!.message).toContain('instantiate');
  });

  it('composeDirect_subnetWithChannel_throwsNamingChannel', () => {
    let err: Error | undefined;
    try {
      PetriNet.builder('Host').compose(retryPolicy());
    } catch (e) {
      err = e as Error;
    }
    expect(err).toBeDefined();
    expect(err!.message).toContain('channel');
    expect(err!.message).toContain('attempt');
    expect(err!.message).toContain('instantiate');
  });

  it('composeDirect_emptySubnet_isNoOp', () => {
    const empty = SubnetDef.builder('Empty').build();
    const host = PetriNet.builder('Host').compose(empty).build();
    expect(host.places.size).toBe(0);
    expect(host.transitions.size).toBe(0);
  });

  it('composeDirect_arclessStandalonePlace_flowsThrough', () => {
    const standalone = place<string>('standalone');
    const subnet = SubnetDef.builder('Standalone').place(standalone).build();
    const host = PetriNet.builder('Host').compose(subnet).build();
    expect([...host.places].some((p) => p.name === 'standalone')).toBe(true);
  });

  it('composeDirect_inhibitorArcPlace_mergesByName', () => {
    const hostSlots = place<string>('slots');
    const host = PetriNet.builder('Host').place(hostSlots).compose(leakyBucket(2)).build();

    const reject = [...host.transitions].find((t) => t.name === 'reject')!;
    expect([...reject.inhibitors].some((arc) => arc.place.name === 'slots')).toBe(true);
    expect([...host.places].filter((p) => p.name === 'slots')).toHaveLength(1);
  });

  it('composeDirect_thenFuse_fusionRunsAtBuild', () => {
    const a = place<string>('a');
    const b = place<string>('b');
    const subnet = SubnetDef.builder('AB')
      .transition(Transition.builder('move').inputs(one(a)).outputs(outPlace(b)).build())
      .build();

    const net = PetriNet.builder('Host')
      .compose(subnet)
      .fuse(FusionSet.of('ab', a, b))
      .build();

    expect([...net.places].some((p) => p.name === 'a')).toBe(true);
    expect([...net.places].some((p) => p.name === 'b')).toBe(false);
  });

  it('composeDirect_producerToConsumer_isBoundedAndReachable', () => {
    // SCG-1: a direct-composed producer→pipe←consumer chain is bounded and
    // the sink is reachable — proves the name-merge wired the two subnets.
    const net = PetriNet.builder('DirectComposed').compose(pipeProducer()).compose(pipeConsumer()).build();
    const result = analyzePipeChain(net);

    expect(result.isComplete).toBe(true);
    expect(result.isGoalLive).toBe(true);
  });

  it('composeDirect_orderIndependentUnderStateClassAnalysis', () => {
    // SCG-2: formal backing for order-independence — both compose orders
    // produce nets with identical reachability and state-class structure.
    const prodThenCons = PetriNet.builder('A').compose(pipeProducer()).compose(pipeConsumer()).build();
    const consThenProd = PetriNet.builder('B').compose(pipeConsumer()).compose(pipeProducer()).build();

    const a = analyzePipeChain(prodThenCons);
    const b = analyzePipeChain(consThenProd);

    expect(a.isGoalLive).toBe(b.isGoalLive);
    expect(a.isComplete).toBe(b.isComplete);
    expect(a.stateClassGraph.size()).toBe(b.stateClassGraph.size());
  });

  function analyzePipeChain(net: PetriNet): LivenessResult {
    const seed = [...net.places].find((p) => p.name === 'seed')!;
    const sink = [...net.places].find((p) => p.name === 'sink')!;
    return TimePetriNetAnalyzer.forNet(bindProducers(net))
      .initialMarking(MarkingState.builder().tokens(seed, 1).build())
      .goalPlaces(sink)
      .maxClasses(100)
      .build()
      .analyze();
  }
});

File: java/src/test/java/org/libpetri/core/ComposeDirectTest.java
package org.libpetri.core;

import org.libpetri.fixtures.StructureOnly;
import org.junit.jupiter.api.Test;
import org.libpetri.analysis.MarkingState;
import org.libpetri.analysis.TimePetriNetAnalyzer;
import org.libpetri.core.Arc.In;
import org.libpetri.core.Arc.Out;
import org.libpetri.fixtures.SubnetFixtures;

import java.util.Set;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Tests for {@link PetriNet.Builder#compose(SubnetDef)} — direct composition
 * per <b>MOD-025</b>.
 *
 * <p>Rule: a {@link SubnetDef} composed directly is integrated <i>without</i>
 * instantiation/prefix-renaming. Body places and transitions keep their
 * original names; places merge into the host purely by {@link Place} equality.
 * Direct composition is order-independent.
 */
class ComposeDirectTest {

    /** A producer whose output place is named {@code pipe} (wires by name). */
    private static SubnetDef<Void> pipeProducer() {
        Place<String> seed = Place.of("seed", String.class);
        Place<String> pipe = Place.of("pipe", String.class);
        return SubnetDef.builder("PipeProducer")
            .place(seed)
            .place(pipe)
            .transition(Transition.builder("emit")
                .inputs(In.one(seed))
                .outputs(Out.place(pipe))
                .build())
            .build();
    }

    /** A consumer whose input place is named {@code pipe} (wires by name). */
    private static SubnetDef<Void> pipeConsumer() {
        Place<String> pipe = Place.of("pipe", String.class);
        Place<String> sink = Place.of("sink", String.class);
        return SubnetDef.builder("PipeConsumer")
            .place(pipe)
            .place(sink)
            .transition(Transition.builder("eat")
                .inputs(In.one(pipe))
                .outputs(Out.place(sink))
                .build())
            .build();
    }

    @Test
    void composeDirect_mergesBodyPlacesByName() {
        // producer()'s body place "output" must merge with the host's equal
        // place; the transition stays un-prefixed (MOD-025 AC1, AC2).
        Place<String> hostOutput = Place.of("output", String.class);

        var host = PetriNet.builder("Host")
            .place(hostOutput)
            .compose(SubnetFixtures.producer())
            .build();

        assertTrue(host.places().contains(hostOutput),
            "body place 'output' must merge with the host place");
        assertEquals(1, host.places().stream().filter(p -> p.name().equals("output")).count(),
            "exactly one 'output' place after the merge");
        assertTrue(host.places().stream().anyMatch(p -> p.name().equals("nextItem")),
            "internal body place 'nextItem' arrives un-prefixed");

        var produce = host.transitions().stream()
            .filter(t -> t.name().equals("produce"))
            .findFirst().orElseThrow();
        assertTrue(produce.outputPlaces().contains(hostOutput),
            "produce's output arc references the host place");
    }

    @Test
    void composeDirect_hostHasNoPreDeclaredPlace_arrivesUnPrefixed() {
        var host = PetriNet.builder("Host")
            .compose(SubnetFixtures.producer())
            .build();

        assertTrue(host.places().stream().anyMatch(p -> p.name().equals("output")),
            "'output' arrives via the produce arc, un-prefixed");
        assertTrue(host.places().stream().anyMatch(p -> p.name().equals("nextItem")),
            "'nextItem' arrives un-prefixed");
    }

    @Test
    void composeDirect_introducesNoPrefixSeparator() {
        // MOD-025 AC2: direct composition never introduces a "prefix/" name.
        var host = PetriNet.builder("Host")
            .compose(SubnetFixtures.producer())
            .compose(SubnetFixtures.consumer())
            .build();

        assertTrue(host.places().stream().noneMatch(p -> p.name().contains("/")),
            "no place name may contain the '/' instance separator");
        assertTrue(host.transitions().stream().noneMatch(t -> t.name().contains("/")),
            "no transition name may contain the '/' instance separator");
    }

    @Test
    void composeDirect_isOrderIndependent() {
        // MOD-025 AC3: the headline property — composing two subnets that
        // share place "pipe" in either order yields the same flat net.
        var prodThenCons = PetriNet.builder("Host")
            .compose(pipeProducer())
            .compose(pipeConsumer())
            .build();
        var consThenProd = PetriNet.builder("Host")
            .compose(pipeConsumer())
            .compose(pipeProducer())
            .build();

        assertEquals(prodThenCons.places(), consThenProd.places(),
            "place sets must be equal regardless of compose order");
        assertEquals(
            prodThenCons.transitions().stream().map(Transition::name).collect(java.util.stream.Collectors.toSet()),
            consThenProd.transitions().stream().map(Transition::name).collect(java.util.stream.Collectors.toSet()),
            "transition name sets must be equal regardless of compose order");

        // Per-transition arc topology must match (Transition equality is
        // identity-based, so compare arc-place projections by transition name).
        for (var name : Set.of("emit", "eat")) {
            var a = prodThenCons.transitions().stream()
                .filter(t -> t.name().equals(name)).findFirst().orElseThrow();
            var b = consThenProd.transitions().stream()
                .filter(t -> t.name().equals(name)).findFirst().orElseThrow();
            assertEquals(
                a.inputSpecs().stream().map(arc -> arc.place()).toList(),
                b.inputSpecs().stream().map(arc -> arc.place()).toList(),
                "input places must match for '" + name + "'");
            assertEquals(a.outputPlaces(), b.outputPlaces(),
                "output places must match for '" + name + "'");
        }
    }

    @Test
    void composeDirect_structurallyEqualToHandWrittenFlatNet() {
        // MOD-025 AC9: a direct-composed net is indistinguishable from a
        // hand-written flat net for the same wiring.
        var composed = PetriNet.builder("Net")
            .compose(pipeProducer())
            .compose(pipeConsumer())
            .build();

        Place<String> seed = Place.of("seed", String.class);
        Place<String> pipe = Place.of("pipe", String.class);
        Place<String> sink = Place.of("sink", String.class);
        var handWritten = PetriNet.builder("Net")
            .transition(Transition.builder("emit")
                .inputs(In.one(seed)).outputs(Out.place(pipe)).build())
            .transition(Transition.builder("eat")
                .inputs(In.one(pipe)).outputs(Out.place(sink)).build())
            .build();

        assertEquals(handWritten.places(), composed.places(),
            "place set must match the hand-written net");
        assertEquals(
            handWritten.transitions().stream().map(Transition::name).collect(java.util.stream.Collectors.toSet()),
            composed.transitions().stream().map(Transition::name).collect(java.util.stream.Collectors.toSet()),
            "transition names must match the hand-written net");
    }

    @Test
    void composeDirect_mixedWithInstanceCompose_coexist() {
        // MOD-025 AC8: direct (un-prefixed) and instance (prefixed) composition
        // coexist in one builder.
        var host = PetriNet.builder("Host")
            .compose(SubnetFixtures.producer())
            .compose(SubnetFixtures.consumer().instantiate("c1"))
            .build();

        var names = host.transitions().stream().map(Transition::name).toList();
        assertTrue(names.contains("produce"),
            "directly composed transition stays un-prefixed. Got: " + names);
        assertTrue(names.contains("c1/consume"),
            "instance-composed transition keeps its prefix. Got: " + names);
    }

    @Test
    void composeDirect_transitionNameCollision_throws() {
        // MOD-025 AC4: composing producer() twice collides on 'produce'.
        var hostBuilder = PetriNet.builder("Host")
            .compose(SubnetFixtures.producer());

        var ex = assertThrows(IllegalStateException.class,
            () -> hostBuilder.compose(SubnetFixtures.producer()));
        var msg = ex.getMessage();
        assertTrue(msg.contains("produce"),
            "error must name the colliding transition 'produce'. Got: " + msg);
        assertTrue(msg.contains("instantiate"),
            "error must point at the instantiate() path. Got: " + msg);
    }

    @Test
    void composeDirect_subnetWithChannel_throwsNamingChannel() {
        // MOD-025 AC5: retryPolicy() declares the channel 'attempt'.
        var hostBuilder = PetriNet.builder("Host");

        var ex = assertThrows(IllegalStateException.class,
            () -> hostBuilder.compose(SubnetFixtures.retryPolicy()));
        var msg = ex.getMessage();
        assertTrue(msg.contains("channel"),
            "error must mention channels. Got: " + msg);
        assertTrue(msg.contains("attempt"),
            "error must name the channel 'attempt'. Got: " + msg);
        assertTrue(msg.contains("instantiate"),
            "error must point at the instantiate() path. Got: " + msg);
    }

    @Test
    void composeDirect_placeTokenTypeConflict_throws() {
        // MOD-025 AC6: host has 'output':Integer, producer() has 'output':String.
        var hostBuilder = PetriNet.builder("Host")
            .place(Place.of("output", Integer.class));

        var ex = assertThrows(IllegalArgumentException.class,
            () -> hostBuilder.compose(SubnetFixtures.producer()));
        var msg = ex.getMessage();
        assertTrue(msg.contains("output"),
            "error must name the conflicting place 'output'. Got: " + msg);
        assertTrue(msg.contains("Integer") && msg.contains("String"),
            "error must name both token types. Got: " + msg);
    }

    @Test
    void composeDirect_emptySubnet_isNoOp() {
        // MOD-025 AC7.
        var empty = SubnetDef.builder("Empty").build();
        var host = PetriNet.builder("Host")
            .compose(empty)
            .build();

        assertTrue(host.places().isEmpty(), "no places from an empty subnet");
        assertTrue(host.transitions().isEmpty(), "no transitions from an empty subnet");
    }

    @Test
    void composeDirect_arclessStandalonePlace_flowsThrough() {
        // A body place referenced by no arc must still appear (places are
        // added explicitly, not only auto-collected from arcs).
        Place<String> standalone = Place.of("standalone", String.class);
        var subnet = SubnetDef.builder("Standalone")
            .place(standalone)
            .build();

        var host = PetriNet.builder("Host")
            .compose(subnet)
            .build();

        assertTrue(host.places().contains(standalone),
            "an arc-less standalone body place must survive direct composition");
    }

    @Test
    void composeDirect_inhibitorArcPlace_mergesByName() {
        // leakyBucket()'s 'reject' transition has an inhibitor on 'slots';
        // direct composition must merge 'slots' with the host's equal place.
        Place<String> hostSlots = Place.of("slots", String.class);
        var host = PetriNet.builder("Host")
            .place(hostSlots)
            .compose(SubnetFixtures.leakyBucket(2))
            .build();

        var reject = host.transitions().stream()
            .filter(t -> t.name().equals("reject"))
            .findFirst().orElseThrow();
        assertTrue(reject.inhibitors().stream().anyMatch(arc -> arc.place().equals(hostSlots)),
            "the inhibitor arc place must merge with the host 'slots' place");
    }

    @Test
    void composeDirect_thenFuse_fusionRunsAtBuild() {
        // Direct composition and fusion interoperate: fuse() runs at build()
        // over the un-prefixed transitions a direct compose contributed.
        Place<String> a = Place.of("a", String.class);
        Place<String> b = Place.of("b", String.class);
        var subnet = SubnetDef.builder("AB")
            .transition(Transition.builder("move")
                .inputs(In.one(a))
                .outputs(Out.place(b))
                .build())
            .build();

        var net = PetriNet.builder("Host")
            .compose(subnet)
            .fuse(FusionSet.of("ab", a, b))
            .build();

        assertTrue(net.places().contains(a), "canonical fused place 'a' survives");
        assertFalse(net.places().contains(b), "non-canonical fused place 'b' is substituted away");
    }

    @Test
    void composeDirect_producerToConsumer_isBoundedAndReachable() {
        // SCG-1: a direct-composed producer→pipe←consumer chain is bounded
        // (finite state class graph) and the sink is reachable — proves the
        // name-merge actually wired the two subnets through 'pipe'.
        var net = PetriNet.builder("DirectComposed")
            .compose(pipeProducer())
            .compose(pipeConsumer())
            .build();

        var seed = net.places().stream()
            .filter(p -> p.name().equals("seed")).findFirst().orElseThrow();
        var sink = net.places().stream()
            .filter(p -> p.name().equals("sink")).findFirst().orElseThrow();

        var result = TimePetriNetAnalyzer.forNet(StructureOnly.bind(net))
            .initialMarking(MarkingState.builder().tokens(seed, 1).build())
            .goalPlaces(sink)
            .maxClasses(100)
            .build()
            .analyze();

        assertTrue(result.isComplete(),
            "direct-composed net must be bounded. Report:\n" + result.report());
        assertTrue(result.isGoalLive(),
            "the sink token must become reachable. Report:\n" + result.report());
    }

    @Test
    void composeDirect_orderIndependentUnderStateClassAnalysis() {
        // SCG-2: formal backing for AC3 — the two compose orders produce nets
        // with identical reachability and identical state-class structure.
        var prodThenCons = PetriNet.builder("A")
            .compose(pipeProducer())
            .compose(pipeConsumer())
            .build();
        var consThenProd = PetriNet.builder("B")
            .compose(pipeConsumer())
            .compose(pipeProducer())
            .build();

        var a = analyzePipeChain(prodThenCons);
        var b = analyzePipeChain(consThenProd);

        assertEquals(a.isGoalLive(), b.isGoalLive(),
            "reachability must not depend on compose order");
        assertEquals(a.isComplete(), b.isComplete(),
            "boundedness must not depend on compose order");
        assertEquals(a.stateClassGraph().size(), b.stateClassGraph().size(),
            "state-class count must not depend on compose order");
    }

    private static TimePetriNetAnalyzer.LivenessResult analyzePipeChain(PetriNet net) {
        var seed = net.places().stream()
            .filter(p -> p.name().equals("seed")).findFirst().orElseThrow();
        var sink = net.places().stream()
            .filter(p -> p.name().equals("sink")).findFirst().orElseThrow();
        return TimePetriNetAnalyzer.forNet(StructureOnly.bind(net))
            .initialMarking(MarkingState.builder().tokens(seed, 1).build())
            .goalPlaces(sink)
            .maxClasses(100)
            .build()
            .analyze();
    }
}
